thespian/system/multiprocCommon.py

```diff
@@ -512,6 +510,69 @@ def shutdown_signal_detected(signum, frame):
     return shutdown_signal_detected
 
 
+def get_min_log_level(logDefs):
+    """
+    Determine the minimum logging level based on the rules in:
+    https://docs.python.org/3/library/logging.config.html#logging.config.dictConfig
+
+    Note: The level and handlers entries are interpreted as for the root logger,
+    except that if a non-root logger’s level is specified as NOTSET, the system
+    consults loggers higher up the hierarchy to determine the effective
+    level of the logger.
+
+    :param logDefs: a logging configuration dictionary object
+    :return: integer value of the lowest log level, or the default level: logging.WARNING.
+
+    >>> import logging
+    >>> logDefs = {'version': 1, 'handlers': {'h1': {'class':
+    ... 'logging.FileHandler', 'filename': 'example.log', 'formatter':
+    ... 'normal', 'filters': ['notActorLog'], 'level': "INFO"}, 'h2': {'class':
+    ... 'logging.FileHandler', 'filename': 'example.log', 'formatter':
+    ... 'actor', 'filters': ['isActorLog'], 'level': "INFO"}},
+    ... 'loggers': {'': {'handlers': ['h1', 'h2'], 'level': "DEBUG"}}}
+    >>> get_min_log_level(logDefs) == logging.DEBUG
+    True
+    >>> logDefs = {'version': 1, 'loggers': {'root' : {'level': 10}, 'level': 20}}
+    >>> min1 = get_min_log_level(logDefs)
+    >>> logDefs = {'version': 1, 'loggers': {'root': {'level': 20}, 'level': 10}}
+    >>> min2 = get_min_log_level(logDefs)
+    >>> min1 == min2 == 10
+    True
+    >>> get_min_log_level({"version" : 1}) == logging.WARNING # default log level
+    True
+    >>> logDefs = {'version': 1, 'loggers': {'root': {'level': "INFO"},
+    ... 'handler2': {"level": logging.NOTSET}}}
+    >>> get_min_log_level(logDefs) == logging.INFO
+    True
+
+    """
+    levels = []
+    names_to_levels = {"NOTSET": logging.NOTSET, "DEBUG": logging.DEBUG,
+                       "INFO": logging.INFO, "WARNING": logging.WARNING,
+                       "ERROR": logging.ERROR, "CRITICAL": logging.CRITICAL}
+
+    loggers = logDefs.get('loggers', {})
+    root_val = loggers.get("", loggers.get("root", None))
+    if root_val:
+        root_val = root_val.get('level')
+    root_val = names_to_levels.get(root_val, root_val)
+
+    def dfs(mapping):
+        """Traverse and collect"""
+        for key, val in mapping.items():
+            if isinstance(val, dict):
+                dfs(val)
+            if key == 'level':
+                if isinstance(val, str):
+                    val = names_to_levels[val.upper()]
+                if val == logging.NOTSET and root_val and root_val > logging.NOTSET:
```

Comment:
It seems to me  that this should exclude all `NOTSET` values, irrespective of `root_val`, since allowing any `NOTSET` would result in a return value of 0 if there was no `root_val` specified.

Also, the code for setting `root_val` does not ensure that `root_val` is numeric: it could be `None`, or it could be an unrecognized string like `"MyLevel"`, which would cause the `>` comparison to throw a `TypeError` exception.

If there's no comparison to `root_val` here, then the reduction would lead to removing `root_val` as a variable and just putting the result of the corresponding expression in the `levels` array.

Let me know if you disagree with my analysis here or if there's something I've missed.

Reply:
NOTSET is useful because the default logging level is warning. If somebody writes some library code and configs the logs somewhere and specifies NOTSET, it does log everything, but if a root logger specifies a level then it supercedes the NOTSET value. But if a logging config had all NOTSET values, and we discarded them, then the configured logging level would default to WARNING, which seems to violate the principle of least surprise.

Also, it looks like root value can be none:
>>> logDefs = {'version': 1, 'root': {'level': 10}, 'loggers': {'one': {'level': 20}}}
>>> logDefs2 = {'version': 1, 'root': None, 'loggers': {'one': {'level': 20}}}
>>> logging.config.dictConfig(logDefs)
>>> logging.config.dictConfig(logDefs2)
>>> 

I like this validation, so I added it to the doctests.

Reply:
I've adjusted the logic. Invalid strings blow up the underlying Python API, so I think that error is fair game.

Reply:
I like those two logDefs as well, although I don't see that you added  them to the doctests.

I agree on the unrecognized strings issue, it's reasonable to have similar behavior.

I'm not seeing behavior where `NOTSET` logs everything.  See https://gist.github.com/kquick/d554a8b4de5541eb48c0bc1b14080fff which is careful to run a separate python for each configuration.  I believe the last output set is the case you indicated should log everything but I see only WARNING and ERROR logging from that output (shown in this gist: https://gist.github.com/kquick/98eac6e6c55fa14fe3fbb19d671ed4b2)

These are the two cases that will still fail:

```
>>> logDefs = {'version': 1,
...             'root': {'level': None},
...            'loggers': {'one': {'level': logging.NOTSET}}}
>>> logging.config.dictConfig(logDefs)
>>> get_min_log_level(logDefs) == logging.WARNING
True

>>> logDefs = {'version': 1,
...            'loggers': {'one': {'level': logging.NOTSET}}}
>>> logging.config.dictConfig(logDefs)
>>> get_min_log_level(logDefs) == logging.WARNING
True
```

Reply:
NOTSET on the root level logs everything, but NOTSET on another level  besides root, logs at the root level, and if unintialized the logging level goes to WARNING.
I added a 6th test case to your gist that I co-opted.
https://gist.github.com/todd-cook/883662ba19eb5a3bb8b372733ec7f6a4

Reply:
see: https://gist.github.com/todd-cook/883662ba19eb5a3bb8b372733ec7f6a4#file-testlog-py-L43-L61
and output at:
https://gist.github.com/todd-cook/883662ba19eb5a3bb8b372733ec7f6a4#file-testlog-py-L114-L135

Reply:
So many different combinations and effects!  The latest `get_min_log_level` in this PR will return the wrong value for case 4 and case 5 if you add those to the doctests.

Also, you can change the doctests from:
```
>>> get_min_log_level(logDevs) == logging.WARNING
True
```
to the following which will provide more information for when the tests fail:
```
>>> get_min_log_level(logDefs)
30
```